tests/subsys/emul/src/emul_test_driver.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "emul_test_driver.h"
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/emul_stub_device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+
+#define DT_DRV_COMPAT vnd_emul_driver
+
+struct emul_test_driver_cfg {
+	int scale;
+};
+
+struct emul_test_driver_data {
+	int action;
+};
+
+static int emul_test_driver_set_action(const struct emul *target, int action)
+{
+	const struct emul_test_driver_cfg *cfg = target->cfg;
+	struct emul_test_driver_data *data = target->data;
+
+	data->action = action * cfg->scale;
+
+	return 0;
+}
+
+static int emul_test_driver_get_action(const struct emul *target, int *action)
+{
+	struct emul_test_driver_data *data = target->data;
+
+	*action = data->action;
+
+	return 0;
+}
+
+static int emul_test_driver_transfer(const struct emul *target, struct i2c_msg *msgs, int num_msgs,
+				     int addr)
+{
+	ARG_UNUSED(target);
+	ARG_UNUSED(msgs);
+	ARG_UNUSED(num_msgs);
+	ARG_UNUSED(addr);
+
+	return -ENOTSUP;
+}
+
+static struct i2c_emul_api bus_api = {
+	.transfer = emul_test_driver_transfer,
+};
+
+static const struct emul_test_backend_api emul_test_driver_backend_api = {
+	.set_action = emul_test_driver_set_action,
+	.get_action = emul_test_driver_get_action,
+};
+
+static int emul_test_driver_init(const struct emul *target, const struct device *parent)
+{
+	ARG_UNUSED(target);
+	ARG_UNUSED(parent);
+
+	return 0;
+}
+
+#define EMUL_TEST_DRIVER(n)                                                                        \
```

Comment:
I can't tell whether this is a structural issue or just semantics, but AFAICT, the point of the stub driver is to satisfy the requirement for a linked `struct device` for `struct emul`s that don't have a natural one. If that's true, then it seems to me like the names of this module and its members should just be "emul_something" and not "something_driver." The test is written in terms of the backend API, which is an emulator API. From the point of view of other modules, this module provides an emulator, and the fact that it contains a driver is an implementation detail.